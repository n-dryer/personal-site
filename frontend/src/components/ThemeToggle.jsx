import React, { useState, useEffect } from 'react';
import { motion } from 'framer-motion';

const ThemeToggle = () => {
  // Initialize with system preference or saved preference
  const [isDark, setIsDark] = useState(() => {
    // Check localStorage first
    const savedTheme = localStorage.getItem('theme');
    if (savedTheme) {
      return savedTheme === 'dark';
    }
    // Otherwise, check system preference
    return window.matchMedia('(prefers-color-scheme: dark)').matches;
  });

  // Apply theme class when component mounts and when theme changes
  useEffect(() => {
    if (isDark) {
      document.documentElement.classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
    }
    // Save preference to localStorage
    localStorage.setItem('theme', isDark ? 'dark' : 'light');
  }, [isDark]);

  const toggleTheme = () => {
    setIsDark(!isDark);
  };

  return (
    <motion.button
      className="theme-toggle"
      onClick={toggleTheme}
      whileTap={{ scale: 0.9 }}
      aria-label={isDark ? "Switch to light mode" : "Switch to dark mode"}
    >
      {isDark ? (
        <svg 
          xmlns="http://www.w3.org/2000/svg" 
          className="h-5 w-5" 
          viewBox="0 0 20 20" 
          fill="currentColor"
        >
          <path 
            fillRule="evenodd" 
            d="M10 2a1 1 0 011 1v1a1 1 0 11-2 0V3a1 1 0 011-1zm4 8a4 4 0 11-8 0 4 4 0 018 0zm-.464 4.95l.707.707a1 1 0 001.414-1.414l-.707-.707a1 1 0 00-1.414 1.414zm2.12-10.607a1 1 0 010 1.414l-.706.707a1 1 0 11-1.414-1.414l.707-.707a1 1 0 011.414 0zM17 11a1 1 0 100-2h-1a1 1 0 100 2h1zm-7 4a1 1 0 011 1v1a1 1 0 11-2 0v-1a1 1 0 011-1zM5.05 6.464A1 1 0 106.465 5.05l-.708-.707a1 1 0 00-1.414 1.414l.707.707zm1.414 8.486l-.707.707a1 1 0 01-1.414-1.414l.707-.707a1 1 0 011.414 1.414zM4 11a1 1 0 100-2H3a1 1 0 000 2h1z" 
            clipRule="evenodd" 
          />
        </svg>
      ) : (
        <svg 
          xmlns="http://www.w3.org/2000/svg" 
          className="h-5 w-5" 
          viewBox="0 0 20 20" 
          fill="currentColor"
        >
          <path d="M17.293 13.293A8 8 0 016.707 2.707a8.001 8.001 0 1010.586 10.586z" />
        </svg>
      )}
    </motion.button>
  );
};

export default ThemeToggle;
