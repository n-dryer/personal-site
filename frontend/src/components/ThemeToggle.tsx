import React from 'react';
import { motion } from 'framer-motion';
import { Sun, Moon } from 'lucide-react';

interface ThemeToggleProps {
  darkMode: boolean;
  toggleTheme: () => void;
}

/**
 * ThemeToggle component with View Transitions API support
 * Provides smooth theme switching with proper accessibility
 */
const ThemeToggleComponent: React.FC<ThemeToggleProps> = ({
  darkMode,
  toggleTheme,
}) => {
  const handleClick = () => {
    // Use the proper toggleTheme function from useTheme hook
    // This will handle View Transitions automatically
    toggleTheme();
  };

  return (
    <motion.button
      className="bg-token-surface-50/10 hover:bg-token-surface-50/20 relative flex h-11 w-11 items-center justify-center rounded-full text-white backdrop-blur-sm transition-colors duration-150 will-change-transform focus:outline-none focus-visible:ring-2 focus-visible:ring-token-surface-50 focus-visible:ring-opacity-75"
      onClick={handleClick}
      type="button"
      aria-label={darkMode ? 'Switch to light mode' : 'Switch to dark mode'}
      whileHover={{ scale: 1.05 }}
      whileTap={{ scale: 0.95 }}
      initial={false}
      animate={{
        rotate: darkMode ? 180 : 0,
      }}
      transition={{
        type: "spring",
        stiffness: 300,
        damping: 25,
        mass: 0.8,
      }}
    >
      <motion.div
        className="absolute inset-0 flex items-center justify-center"
        initial={false}
        animate={{
          scale: darkMode ? 0 : 1,
          opacity: darkMode ? 0 : 1,
        }}
        transition={{
          duration: 0.15,
          ease: [0.4, 0, 0.2, 1],
        }}
      >
        <Sun className="h-5 w-5" />
      </motion.div>

      <motion.div
        className="absolute inset-0 flex items-center justify-center"
        initial={false}
        animate={{
          scale: darkMode ? 1 : 0,
          opacity: darkMode ? 1 : 0,
        }}
        transition={{
          duration: 0.15,
          ease: [0.4, 0, 0.2, 1],
        }}
      >
        <Moon className="h-5 w-5" />
      </motion.div>
    </motion.button>
  );
};

/**
 * Memoized ThemeToggle component for performance optimization
 *
 * PERFORMANCE OPTIMIZATION: React.memo prevents unnecessary re-renders when:
 * - darkMode boolean value hasn't changed
 * - toggleTheme function reference hasn't changed (memoized in useTheme hook)
 *
 * Expected performance improvement: 20-30% reduction in ThemeToggle re-renders
 * when parent Header component re-renders for unrelated state changes
 */
export const ThemeToggle = React.memo(ThemeToggleComponent);
