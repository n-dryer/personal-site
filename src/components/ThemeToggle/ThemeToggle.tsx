import React from 'react';
import { Sun, Moon } from 'lucide-react';
import { motion, AnimatePresence, easeInOut } from 'framer-motion';
import { useViewTransitions } from '../../hooks/useViewTransitions';
import { useReducedMotion } from '../../hooks/useReducedMotion';

/**
 * Props for the ThemeToggle component.
 */
type ThemeToggleProps = {
  /** Indicates if dark mode is currently active. */
  darkMode: boolean;
  /** Function to toggle the theme. */
  toggleTheme: () => void;
};

/**
 * ThemeToggleComponent is a button that allows users to switch between light and dark themes.
 * It displays a sun icon for light mode and a moon icon for dark mode, with animations.
 * The component uses view transitions if supported by the browser.
 * @param {ThemeToggleProps} props - The properties passed to the component.
 * @returns {JSX.Element} The rendered ThemeToggleComponent.
 */
const ThemeToggleComponent = ({ darkMode, toggleTheme }: ThemeToggleProps) => {
  const { withViewTransition } = useViewTransitions();
  const shouldReduceMotion = useReducedMotion();

  const handleClick = () => {
    withViewTransition(toggleTheme);
  };

  const iconVariants = {
    initial: { opacity: 1, rotate: -90, scale: 0.5 },
    animate: { opacity: 1, rotate: 0, scale: 1 },
    exit: { opacity: 1, rotate: 90, scale: 0.5 },
  };

  return (
    <button
      className="bg-surface/80 hover:bg-surface/90 relative flex h-14 w-14 items-center justify-center rounded-full border border-white/10 text-text-secondary ring-1 ring-white/5 backdrop-blur-xl transition-colors duration-150 focus:outline-none focus-visible:ring-2 focus-visible:ring-accent focus-visible:ring-opacity-75"
      onClick={handleClick}
      type="button"
      aria-label={darkMode ? 'Switch to light mode' : 'Switch to dark mode'}
    >
      <div className="relative">
        <AnimatePresence initial={false} mode="wait">
          {darkMode ? (
            <motion.div
              key="moon"
              variants={shouldReduceMotion ? undefined : iconVariants}
              initial="initial"
              animate="animate"
              exit="exit"
              transition={{ duration: 0.25, ease: easeInOut }}
              style={{ viewTransitionName: 'theme-icon-moon' }}
            >
              <Moon className="h-8 w-8 text-accent" />
            </motion.div>
          ) : (
            <motion.div
              key="sun"
              variants={shouldReduceMotion ? undefined : iconVariants}
              initial="initial"
              animate="animate"
              exit="exit"
              transition={{ duration: 0.25, ease: easeInOut }}
              style={{ viewTransitionName: 'theme-icon-sun' }}
            >
              <Sun className="h-8 w-8 text-accent" />
            </motion.div>
          )}
        </AnimatePresence>
      </div>
    </button>
  );
};

/**
 * Memoized ThemeToggle component for performance optimization.
 * This component allows users to switch between light and dark themes.
 * @see ThemeToggleComponent
 */
export const ThemeToggle = React.memo(ThemeToggleComponent);
