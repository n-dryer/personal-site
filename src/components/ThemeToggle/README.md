# ThemeToggle Component

This directory contains the `ThemeToggle` component, its tests, and related files.

## Purpose

The `ThemeToggle` component is a button that allows users to switch between light and dark themes for the application. It displays a sun icon for light mode and a moon icon for dark mode, with smooth animations and respects user's reduced motion preferences.

## Usage

```tsx
import { ThemeToggle } from '@/components/ThemeToggle'; // Or the correct relative path

const MyComponent = () => {
  const [darkMode, setDarkMode] = React.useState(false);
  const handleToggleTheme = () => setDarkMode(prev => !prev);

  return <ThemeToggle darkMode={darkMode} toggleTheme={handleToggleTheme} />;
}
```

## Props

- `darkMode`: `boolean` - Indicates whether dark mode is currently active.
- `toggleTheme`: `() => void` - A callback function that is executed when the button is clicked, responsible for changing the theme state. 