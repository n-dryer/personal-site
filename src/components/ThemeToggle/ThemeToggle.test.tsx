import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { ThemeToggle } from './ThemeToggle'; // Adjusted import path

// Mock framer-motion - only need motion.div for the icon transitions
jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: React.PropsWithChildren<any>) => (
      <div {...props}>{children}</div> // Basic div mock for icons
    ),
  },
  AnimatePresence: ({ children }: React.PropsWithChildren<any>) => (
    <>{children}</> // Pass through children
  ),
}));

// Mock lucide-react icons
jest.mock('lucide-react', () => ({
  Sun: () => <div data-testid='sun-icon'>Sun Icon</div>,
  Moon: () => <div data-testid='moon-icon'>Moon Icon</div>,
}));

// Mock useViewTransitions and useReducedMotion hooks
jest.mock('../../hooks/useViewTransitions', () => ({ // Adjusted path
  useViewTransitions: () => ({
    withViewTransition: (callback: () => void) => callback(), // Directly call callback
  }),
}));

jest.mock('../../hooks/useReducedMotion', () => ({ // Adjusted path
  useReducedMotion: () => false, // Assume motion is not reduced for these tests
}));

describe('ThemeToggle Component', () => {
  const mockToggleTheme = jest.fn();

  beforeEach(() => {
    mockToggleTheme.mockClear();
    // localStorage is not directly used by ThemeToggle, but by the useTheme hook it would normally consume.
    // For this component test, direct interaction is primary.
  });

  it('should render the correct icon and aria-label based on the darkMode prop', () => {
    const { rerender } = render(
      <ThemeToggle darkMode={false} toggleTheme={mockToggleTheme} />
    );
    // Light mode
    expect(screen.getByLabelText('Switch to dark mode')).toBeInTheDocument();
    expect(screen.getByTestId('sun-icon')).toBeInTheDocument();
    expect(screen.queryByTestId('moon-icon')).not.toBeInTheDocument();

    // Dark mode
    rerender(<ThemeToggle darkMode={true} toggleTheme={mockToggleTheme} />);
    expect(screen.getByLabelText('Switch to light mode')).toBeInTheDocument();
    expect(screen.getByTestId('moon-icon')).toBeInTheDocument();
    expect(screen.queryByTestId('sun-icon')).not.toBeInTheDocument();
  });

  it('should call toggleTheme when the button is clicked', () => {
    render(<ThemeToggle darkMode={false} toggleTheme={mockToggleTheme} />);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(mockToggleTheme).toHaveBeenCalledTimes(1);
  });
}); 