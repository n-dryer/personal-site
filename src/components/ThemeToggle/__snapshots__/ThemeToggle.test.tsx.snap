// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeToggle Component renders correctly 1`] = `
<div>
  <button
    animate="[object Object]"
    aria-label="Switch to dark mode"
    class="relative flex h-11 w-11 items-center justify-center rounded-full bg-white/10 text-white backdrop-blur-sm transition-all duration-200 hover:bg-white/20 focus:outline-none focus-visible:ring-2 focus-visible:ring-white focus-visible:ring-opacity-75 active:scale-95"
    data-testid="motion-button"
    transition="[object Object]"
    type="button"
    whilehover="[object Object]"
    whiletap="[object Object]"
  >
    <div
      animate="[object Object]"
      class="absolute inset-0 flex items-center justify-center"
      data-testid="motion-div"
      transition="[object Object]"
    >
      <div
        data-testid="sun-icon"
      >
        Sun Icon
      </div>
    </div>
    <div
      animate="[object Object]"
      class="absolute inset-0 flex items-center justify-center"
      data-testid="motion-div"
      transition="[object Object]"
    >
      <div
        data-testid="moon-icon"
      >
        Moon Icon
      </div>
    </div>
  </button>
</div>
`;
